src/main/java/com/empowerops/linqalike/queries/CastQuery.java

```diff
@@ -1,26 +1,75 @@
 package com.empowerops.linqalike.queries;
 
+import com.empowerops.linqalike.DefaultedBiQueryable;
 import com.empowerops.linqalike.DefaultedQueryable;
+import com.empowerops.linqalike.common.Tuple;
 
 import java.util.Iterator;
 import java.util.NoSuchElementException;
+import java.util.Optional;
 
-public class CastQuery<TUncasted, TCasted> implements DefaultedQueryable<TCasted>, FastSize {
+public class CastQuery<TUncasted, TCast> implements DefaultedQueryable<TCast>, FastSize {
 
     private final Iterable<TUncasted> sourceElements;
-    private final Class<TCasted> desiredType;
+    private final Class<TCast> desiredType;
 
-    public CastQuery(Iterable<TUncasted> sourceElements, Class<TCasted> desiredType) {
+    public CastQuery(Iterable<TUncasted> sourceElements, Class<TCast> desiredType) {
         this.sourceElements = sourceElements;
         this.desiredType = desiredType;
     }
 
     @Override
-    public Iterator<TCasted> iterator() {
+    public Iterator<TCast> iterator() {
         return new CastIterator();
     }
 
-    private class CastIterator implements Iterator<TCasted>{
+    public static class Inner<TLeftCast, TLeftOriginal, TRightCast, TRightOriginal>
+            implements DefaultedBiQueryable<TLeftCast, TRightCast> {
+
+        private final Iterable<Tuple<TLeftOriginal, TRightOriginal>> sourceElements;
+        private final Optional<Class<TLeftCast>> desiredLeftType;
+        private final Optional<Class<TRightCast>> desiredRightType;
+
+        public Inner(Iterable<Tuple<TLeftOriginal, TRightOriginal>> sourceElements,
+                     Optional<Class<TLeftCast>> desiredLeftType,
+                     Optional<Class<TRightCast>> desiredRightType) {
+
+            assert desiredLeftType.isPresent() || desiredRightType.isPresent() : "no cast to perform?!";
+
+            this.sourceElements = sourceElements;
+            this.desiredLeftType = desiredLeftType;
+            this.desiredRightType = desiredRightType;
+        }
+
+        @Override
+        public Iterator<Tuple<TLeftCast, TRightCast>> iterator() {
+            return new CastLeftIterator();
+        }
+
+        public class CastLeftIterator implements Iterator<Tuple<TLeftCast, TRightCast>> {
+
+            private final Iterator<Tuple<TLeftOriginal, TRightOriginal>> sourceIterator = sourceElements.iterator();
+
+            @Override
+            public boolean hasNext() {
+                return sourceIterator.hasNext();
+            }
+
+            @SuppressWarnings("unchecked") //type checked dynamically
+            @Override
+            public Tuple<TLeftCast, TRightCast> next() {
+                Tuple<TLeftOriginal, TRightOriginal> next = sourceIterator.next();
+                // nothing to be gained from using new memory here, since Tuple is immutable,
+                // and java doesn't support casting operators
+                // so we'll do a type-check and return the original
```

Comment:
this can be re-worded 

> nothing to be gained in using new memory on a static type here (a new Tuple<>)
> and java doesn't support overriding casting operators
> so we'll do a type check and return the original